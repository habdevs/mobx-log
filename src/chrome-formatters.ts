import { isObservableArray, isObservableObject, toJS } from 'mobx';

type Css = { style: string };
type Chunk = [string, Css, string];
type Header = [string, Chunk, [string, string]];
type Body = [string, Css, ...any[]];

export type ChromeFormatter<T> = {
  header: (argument: T) => Header | null;
  hasBody: (argument: T) => boolean;
  body: (argument: T) => Body;
};

const styles = {
  mobxName: { style: 'color: rgb(232,98,0)' },
  nullName: { style: 'color: #777' },
  list: {
    style:
      'list-style-type: none; padding: 0; margin: 0 0 0 12px; font-style: normal',
  },
};

const reference = (object: unknown) => {
  if (typeof object === 'undefined' || typeof object === null) {
    return ['span', styles.nullName, typeof object];
  }
  return ['object', { object }];
};

const renderIterableHeader = (
  iterable: ArrayLike<unknown>,
  name: string
): Header => [
  'span',
  ['span', styles.mobxName, name],
  ['span', `[${iterable.length}]`],
];

const hasBody = (collection: ArrayLike<unknown>) => collection.length > 0;

const renderIterableBody = (
  collection: ArrayLike<unknown> | {},
  mapper: (val: [string, unknown]) => unknown
): Body => {
  const children = Object.entries(toJS(collection)).map(mapper);
  return ['ol', styles.list, ...children];
};

export class ObjectFormatter implements ChromeFormatter<{}> {
  header(argument: {}): Header | null {
    return isObservableObject(argument)
      ? renderIterableHeader(Object.keys(argument), 'Object')
      : null;
  }

  hasBody(argument: {}): boolean {
    return hasBody(Object.keys(argument));
  }

  body(argument: {}) {
    return renderIterableBody(argument, ([key, value]) => [
      'li',
      {},
      reference(key),
      ': ',
      reference(value),
    ]);
  }
}

export class ArrayFormatter implements ChromeFormatter<ArrayLike<unknown>> {
  body(argument: ArrayLike<unknown>): Body {
    return renderIterableBody(argument, ([_, value]) => [
      'li',
      reference(value),
    ]);
  }

  hasBody(argument: ArrayLike<unknown>): boolean {
    return hasBody(argument);
  }

  header(argument: ArrayLike<unknown>): Header | null {
    return isObservableArray(argument)
      ? renderIterableHeader(argument, 'Array')
      : null;
  }
}
