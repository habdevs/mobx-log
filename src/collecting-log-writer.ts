import { LogWriter } from './types';

export class CollectingLogWriter implements LogWriter {
  history: string[] = [];

  write(...messages: unknown[]) {
    const strings = messages.map((message): string =>
      (message as any).toString()
    );
    this.history.push(strings.join(' '));
  }
}
