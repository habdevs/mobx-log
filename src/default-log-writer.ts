import { LogWriter } from './types';

export class DefaultLogWriter implements LogWriter {
  write(...messages: unknown[]) {
    console.log(...messages);
  }
}
