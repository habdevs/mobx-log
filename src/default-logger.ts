import { ActionEvent, ComputedEvent, Logger, ObservableEvent } from "./logger";

type Log = (...messages: unknown[]) => void;
type Now = () => string;

export class DefaultLogger implements Logger {
  debug = false;

  constructor(private log: Log, private now: Now) {}

  logObservable(event: ObservableEvent): void {
    if (this.debug) {
      this.log(event);
    }
    this.log(
      this.now(),
      "OBSERVABLE",
      event.name,
      event.oldValue,
      "->",
      event.newValue
    );
  }

  logAction(event: ActionEvent): void {
    if (this.debug) {
      this.log(event);
    }
    this.log(this.now(), "ACTION", event.name, event.arguments);
  }

  logComputed(event: ComputedEvent): void {
    if (this.debug) {
      this.log(event);
    }
    console.log(
      this.now(),
      "COMPUTED",
      event.name,
      event.oldValue,
      "->",
      event.newValue
    );
  }
}
