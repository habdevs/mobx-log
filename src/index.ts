export { enableLogging } from "./enable-logging";
