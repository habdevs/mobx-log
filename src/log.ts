import type { PureSpyEvent } from "mobx/src/core/spy";
import { IEqualsComparer } from "mobx";
import { Logger } from "./logger";

type SpyEvent = PureSpyEvent & {
  object: any;
  name: string;
};

export type LogOptions = {
  logger: Logger;
  compare: IEqualsComparer<unknown>;
};

export const log = (event: SpyEvent, options: LogOptions): void => {
  const { logger, compare } = options;

  if (event.type === "update") {
    if (event.observableKind === "computed") {
      if (!compare(event.oldValue, event.newValue)) {
        logger.logComputed({
          name: event.debugObjectName,
          oldValue: event.oldValue,
          newValue: event.newValue,
        });
      }
    }
    if (event.observableKind === "object") {
      logger.logObservable({
        name: `${event.debugObjectName}.${event.name}`,
        newValue: event.newValue,
        oldValue: event.oldValue,
      });
    }
  }
  if (event.type === "action") {
    logger.logAction({
      name: `${event.object.constructor.name}.${event.name}`,
      arguments: event.arguments,
    });
  }
};
