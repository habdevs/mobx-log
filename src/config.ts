import { DefaultLogger } from './default-logger';
import { Logger } from './types';
import { DefaultLogWriter } from './log-writer';

type Config = {
  logger: Logger;
};

export const config: Config = {
  logger: new DefaultLogger(new DefaultLogWriter()),
};

export const configureMakeLoggable = (options: Config) => {
  config.logger = options.logger;
};
