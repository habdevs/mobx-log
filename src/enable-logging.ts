import { comparer, spy } from "mobx";
import { log, LogOptions } from "./log";
import { DefaultLogger } from "./default-logger";
import { now } from "./time";

const defaultOptions = {
  logger: new DefaultLogger(console.log, now),
  compare: comparer.default,
};

export const enableLogging = (options: LogOptions = defaultOptions) => {
  // Mobx typings aren't up-to-date
  // @ts-ignore
  spy((event) => log(event, options));
};
