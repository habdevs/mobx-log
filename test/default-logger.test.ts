import { DefaultLogger } from '../src/default-logger';
import { CollectingLogWriter } from '../src/collecting-log-writer';

describe('DefaultLogger', () => {
  it('logs observable', () => {
    const logWriter = new CollectingLogWriter();
    const plainLogger = new DefaultLogger(logWriter, () => '14:34:57');

    plainLogger.logObservable({
      name: 'age',
      oldValue: 23,
      newValue: 24,
    });

    expect(logWriter.history).toMatchSnapshot();

    plainLogger.debug = true;

    plainLogger.logObservable({
      name: 'age',
      oldValue: 24,
      newValue: 25,
    });

    expect(logWriter.history).toMatchSnapshot();
  });

  it('logs computed', () => {
    const logWriter = new CollectingLogWriter();
    const plainLogger = new DefaultLogger(logWriter, () => '14:34:57');

    plainLogger.logComputed({
      name: 'isEven',
      oldValue: true,
      newValue: false,
    });

    expect(logWriter.history).toMatchSnapshot();
  });

  it('logs action', () => {
    const logWriter = new CollectingLogWriter();
    const plainLogger = new DefaultLogger(logWriter, () => '14:34:57');

    plainLogger.logAction({
      name: 'add',
      arguments: [1],
    });

    expect(logWriter.history).toMatchSnapshot();
  });
});
