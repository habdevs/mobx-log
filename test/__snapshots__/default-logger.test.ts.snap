// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultLogger logs observable 1`] = `
Array [
  Array [
    "14:34:57",
    "OBSERVABLE",
    "age",
    23,
    "->",
    24,
  ],
]
`;

exports[`DefaultLogger logs observable 2`] = `
Array [
  Array [
    "14:34:57",
    "OBSERVABLE",
    "age",
    23,
    "->",
    24,
  ],
  Array [
    Object {
      "name": "age",
      "newValue": 25,
      "oldValue": 24,
    },
  ],
  Array [
    "14:34:57",
    "OBSERVABLE",
    "age",
    24,
    "->",
    25,
  ],
]
`;
